File: packages/typescript/src/core/metadata.ts
import { canonicalize, hashBytes } from '@constellation-network/metagraph-sdk';
import type { FingerprintValue, FingerprintMetadata } from './types';

/**
 * Compute the SHA-256 hash of a FingerprintValue's canonical JSON.
 *
 * This follows the same process the server uses to produce the metadata hash:
 * 1. Canonicalize the FingerprintValue object per RFC 8785
 * 2. Encode the canonical string as UTF-8 bytes
 * 3. SHA-256 hash those bytes
 *
 * @param value - The FingerprintValue to hash
 * @returns Hex-encoded SHA-256 hash (64 characters)
 */
export function computeMetadataHash(value: FingerprintValue): string {
  const canonical = canonicalize(value);
  const bytes = new TextEncoder().encode(canonical);
  return hashBytes(bytes).value;
}

/**
 * Create a FingerprintMetadata object with the computed hash and optional tags.
 *
 * @param value - The FingerprintValue to derive the hash from
 * @param tags  - Optional key-value tags for indexing (max 6 pairs per proto)
 * @returns FingerprintMetadata ready for inclusion in a FingerprintSubmission
 */
export function createMetadata(
  value: FingerprintValue,
  tags?: Record<string, string>
): FingerprintMetadata {
  return {
    hash: computeMetadataHash(value),
    ...(tags && { tags }),
  };
}

File: packages/python/src/constellation_ded_sdk/core/metadata.py
"""Metadata construction and hashing."""

from constellation_sdk import canonicalize, hash_bytes

from .types import FingerprintMetadata, FingerprintValue


def compute_metadata_hash(value: FingerprintValue) -> str:
    """Compute the SHA-256 hash of a FingerprintValue's canonical JSON.

    Args:
        value: The FingerprintValue to hash.

    Returns:
        Hex-encoded SHA-256 hash (64 characters).
    """
    canonical = canonicalize(value.to_dict())
    return hash_bytes(canonical.encode("utf-8")).value


def create_metadata(
    value: FingerprintValue,
    tags: dict[str, str] | None = None,
) -> FingerprintMetadata:
    """Create a FingerprintMetadata with computed hash and optional tags.

    Args:
        value: The FingerprintValue to derive the hash from.
        tags: Optional key-value tags for indexing (max 6 pairs per proto).

    Returns:
        FingerprintMetadata ready for inclusion in a FingerprintSubmission.
    """
    return FingerprintMetadata(
        hash=compute_metadata_hash(value),
        tags=tags,
    )
